The reasoner doesn't work with 'high' reasoning rates

When I set the reasoning period to be 1 or less, I get a lot of bugs.

An example:

```
[pipeline_inspection_reasoner-1] [ERROR] [1673354688.878945832] [pipeline_inspection_reasoner]: Error raised in execute callback: list index out of range
[pipeline_inspection_reasoner-1] Traceback (most recent call last):
[pipeline_inspection_reasoner-1]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/action/server.py", line 324, in _execute_goal
[pipeline_inspection_reasoner-1]     execute_result = await await_or_execute(execute_callback, goal_handle)
[pipeline_inspection_reasoner-1]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 107, in await_or_execute
[pipeline_inspection_reasoner-1]     return callback(*args)
[pipeline_inspection_reasoner-1]   File "/home/gus/pipeline_ws/build/mros2_reasoner/mros2_reasoner/ros_reasoner.py", line 156, in objective_action_callback
[pipeline_inspection_reasoner-1]     fg_instance = self.onto.search_one(solvesO=objective)
[pipeline_inspection_reasoner-1]   File "/home/gus/.local/lib/python3.10/site-packages/owlready2/namespace.py", line 395, in search_one
[pipeline_inspection_reasoner-1]     def search_one(self, **kargs): return self.search(**kargs).first()
[pipeline_inspection_reasoner-1]   File "/home/gus/.local/lib/python3.10/site-packages/owlready2/util.py", line 62, in first
[pipeline_inspection_reasoner-1]     if len(self) != 0: return self[0]
[pipeline_inspection_reasoner-1]   File "/home/gus/.local/lib/python3.10/site-packages/owlready2/util.py", line 192, in __getitem__
[pipeline_inspection_reasoner-1]     return self[i]
[pipeline_inspection_reasoner-1] IndexError: list index out of range
[pipeline_inspection_reasoner-1] [WARN] [1673354688.880966255] [pipeline_inspection_reasoner]: Goal state not set, assuming aborted. Goal ID: [ 77  12 182 243  76 147  77 121 161  64   2  40  60 160  33  37]

```

Or 
```
[pipeline_inspection_reasoner-1] [ERROR] [1673533577.337098677] [pipeline_inspection_reasoner]: Error in perform_reasoning: 'NoneType' object is not subscriptable
```

I will add the other error messages here later

> [pipeline_inspection_reasoner-1] [ERROR] [1673533577.337098677] [pipeline_inspection_reasoner]: Error in perform_reasoning: 'NoneType' object is not subscriptable

This error is also happening with lower rates, but less frequently. Unfortunately, I don't have any more info.
My guess is that is has something to do with reading/writing from/to the ontology.

--- a/mros2_reasoner/mros2_reasoner/reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/reasoner.py
@@ -44,18 +44,20 @@ class Reasoner:
 
     def remove_objective(self, objective_id):
         # Checks if there are previously defined objectives.
-        old_objective = self.tomasys.search_one(iri="*{}".format(objective_id))
-        if old_objective:
-            old_fg_instance = self.tomasys.search_one(solvesO=old_objective)
-            with self.ontology_lock:
+        with self.ontology_lock:
+            old_objective = self.tomasys.search_one(
+                iri="*{}".format(objective_id))
+            if old_objective:
+                old_fg_instance = self.tomasys.search_one(
+                    solvesO=old_objective)
                 destroy_entity(old_fg_instance)
                 destroy_entity(old_objective)
-            return True
-        else:
+                return True
             return False
 
     def search_objectives(self):
-        objectives = self.tomasys.search(type=self.tomasys.Objective)
+        with self.ontology_lock:
+            objectives = self.tomasys.search(type=self.tomasys.Objective)
         return objectives
 
     def has_objective(self):
@@ -77,8 +79,7 @@ class Reasoner:
         objective = self.get_objective_from_objective_id(objective_id)
         if objective is None:
             return None
-        else:
-            return str(objective.typeF.name)
+        return str(objective.typeF.name)
 
     def get_objectives_in_error(self):
         return get_objectives_in_error(self.search_objectives())
@@ -99,12 +100,10 @@ class Reasoner:
         """Creates QAvalue individual in the KB given a desired name and a
         string seed for the QAtype name and the value
         """
-
-        # TODO: this search is not optimal, the search + loop can be
-        # substituted by a single search
-        qa_type = self.get_qa_type(nfr_key)
-
         with self.ontology_lock:
+            # TODO: this search is not optimal, the search + loop can be
+            # substituted by a single search
+            qa_type = self.get_qa_type(nfr_key)
             new_nfr = self.tomasys.QAvalue(
                 str(qa_value_name),
                 namespace=self.tomasys,
@@ -113,6 +112,20 @@ class Reasoner:
 
             return new_nfr
 
+    def append_nfr_to_objective(self, objective, nfr):
+        with self.ontology_lock:
+            objective.hasNFR.append(nfr)
+        return objective
+
+    def update_objective_status(self, objective, status):
+        with self.ontology_lock:
+            objective.o_status = status
+        return objective
+
+    def get_fg_solves_objective(self, objective):
+        with self.ontology_lock:
+            return self.tomasys.search_one(solvesO=objective)
+
     def set_new_grounding(self, fd_name, objective):
         """Given a string fd_name with the name of a FunctionDesign and an
         objective, removes the previous fg for the objective and ground a new
@@ -120,15 +133,14 @@ class Reasoner:
         """
         with self.ontology_lock:
             remove_objective_grounding(objective, self.tomasys)
-        fd = self.tomasys.search_one(
-            iri="*{}".format(fd_name),
-            is_a=self.tomasys.FunctionDesign)
-        if fd:
-            with self.ontology_lock:
+            fd = self.tomasys.search_one(
+                iri="*{}".format(fd_name),
+                is_a=self.tomasys.FunctionDesign)
+            if fd:
+                # with self.ontology_lock:
                 ground_fd(fd, objective, self.tomasys)
                 reset_objective_status(objective)
-            return str(fd.name)
-        else:
+                return str(fd.name)
             return None
 
     # the DiagnosticStatus message process contains, per field
@@ -154,19 +166,19 @@ class Reasoner:
     def update_component_status(self, diagnostic_status):
         # Find the Component with the same name that the one in the Component
         # Status message (in diagnostic_status.key)
-        component_type = self.tomasys.search_one(
-            iri="*{}".format(diagnostic_status.values[0].key))
-        if component_type is not None:
-            value = diagnostic_status.values[0].value
-            with self.ontology_lock:
+        with self.ontology_lock:
+            component_type = self.tomasys.search_one(
+                iri="*{}".format(diagnostic_status.values[0].key))
+            if component_type is not None:
+                value = diagnostic_status.values[0].value
                 reset_fd_realisability(
                     self.tomasys,
                     diagnostic_status.values[0].key)
                 component_type.c_status = value
-            return_value = 1
-        else:
-            return_value = 0
-        return return_value
+                return_value = 1
+            else:
+                return_value = 0
+            return return_value
 
     def get_qa_type(self, key):
         # TODO: this search is not optimal, the search + loop can be
@@ -193,19 +205,19 @@ class Reasoner:
 
     # update QA value based on incoming diagnostic
     def update_qa(self, diagnostic_status):
-        qa_type = self.get_qa_type(diagnostic_status.values[0].key)
-        if qa_type is not None:
-            fgs = self.get_function_groudings_require_qa(qa_type)
-            value = float(diagnostic_status.values[0].value)
-            with self.ontology_lock:
+        with self.ontology_lock:
+            qa_type = self.get_qa_type(diagnostic_status.values[0].key)
+            if qa_type is not None:
+                fgs = self.get_function_groudings_require_qa(qa_type)
+                value = float(diagnostic_status.values[0].value)
                 measured_qa = update_measured_qa_value(
                     qa_type, value, self.tomasys)
                 for fg in fgs:
                     update_fg_measured_qa(fg, measured_qa)
-            return_value = True
-        else:
-            return_value = False
-        return return_value
+                return_value = True
+            else:
+                return_value = False
+            return return_value
 
     # EXEC REASONING to update ontology with inferences
     # TODO CHECK: update reasoner facts, evaluate, retrieve action, publish
@@ -232,12 +244,12 @@ class Reasoner:
         sys.exit(0)
 
     def handle_updatable_objectives(self, obj_in_error):
-        if obj_in_error.o_status == "UPDATABLE":
-            self.logger.info(
-                ">> UPDATABLE objective - Try to clear Components status")
-            for comp_inst in list(
-                    self.tomasys.ComponentState.instances()):
-                with self.ontology_lock:
+        with self.ontology_lock:
+            if obj_in_error.o_status == "UPDATABLE":
+                self.logger.info(
+                    ">> UPDATABLE objective - Try to clear Components status")
+                for comp_inst in list(
+                        self.tomasys.ComponentState.instances()):
                     if comp_inst.c_status == "RECOVERED":
                         self.logger.info(
                             "Component {0} Status {1} - Set to None".format(
@@ -258,14 +270,16 @@ class Reasoner:
 
     # find best fds for all objectives
     def select_desired_configuration(self, obj_in_error):
-        self.logger.info(" >> Reasoner searches an FD ")
-        desired_configuration = obtain_best_function_design(
-            obj_in_error, self.tomasys)
+        with self.ontology_lock:
+            self.logger.info(" >> Reasoner searches an FD ")
+            desired_configuration = obtain_best_function_design(
+                obj_in_error, self.tomasys)
 
-        if desired_configuration is None:
-            self.logger.warning(
-                "No FD found to solve Objective {} ".format(obj_in_error.name))
-        return desired_configuration
+            if desired_configuration is None:
+                self.logger.warning(
+                    "No FD found to solve Objective {} ".format(
+                        obj_in_error.name))
+            return desired_configuration
 
     # MAPE-K: Analyze step
     def analyze(self):

--- a/mros2_reasoner/mros2_reasoner/reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/reasoner.py
@@ -25,6 +25,8 @@ from owlready2 import sync_reasoner_pellet
 
 import logging
 
+from typing import Tuple
+
 
 class Reasoner:
 
@@ -42,11 +44,11 @@ class Reasoner:
         self.logger = logging
         signal.signal(signal.SIGINT, self.save_ontology_exit)
 
-    def remove_objective(self, objective_id):
+    def remove_objective(self, objective_name: str) -> bool:
         # Checks if there are previously defined objectives.
         with self.ontology_lock:
             old_objective = self.tomasys.search_one(
-                iri="*{}".format(objective_id))
+                iri="*{}".format(objective_name), is_a=self.tomasys.Objective)
             if old_objective:
                 old_fg_instance = self.tomasys.search_one(
                     solvesO=old_objective)
@@ -67,21 +69,22 @@ class Reasoner:
             has_objective = True
         return has_objective
 
-    def get_objective_from_objective_id(self, objective_id):
+    def get_objective_from_objective_id(self, objective_id: str):
         objectives = self.search_objectives()
         if objectives == []:
             return None
         for objective in objectives:
             if str(objective.name) == str(objective_id):
                 return objective
+        return None
 
-    def get_function_name_from_objective_id(self, objective_id):
+    def get_function_name_from_objective_id(self, objective_id: str) -> str:
         objective = self.get_objective_from_objective_id(objective_id)
         if objective is None:
             return None
         return str(objective.typeF.name)
 
-    def get_objectives_in_error(self):
+    def get_objectives_in_error(self) -> Tuple[list[str], list[str]]:
         return get_objectives_in_error(self.search_objectives())
 
     def get_new_tomasys_objective(self, objective_name, iri_seed):
@@ -126,21 +129,21 @@ class Reasoner:
         with self.ontology_lock:
             return self.tomasys.search_one(solvesO=objective)
 
-    def set_new_grounding(self, fd_name, objective):
+    def set_new_grounding(self, fd_name: str, obj_name: str) -> str | None:
         """Given a string fd_name with the name of a FunctionDesign and an
         objective, removes the previous fg for the objective and ground a new
         fg of typeF fd
         """
         with self.ontology_lock:
-            remove_objective_grounding(objective, self.tomasys)
+            remove_objective_grounding(obj_name, self.tomasys)
             fd = self.tomasys.search_one(
                 iri="*{}".format(fd_name),
                 is_a=self.tomasys.FunctionDesign)
             if fd:
                 # with self.ontology_lock:
-                ground_fd(fd, objective, self.tomasys)
-                reset_objective_status(objective)
-                return str(fd.name)
+                ground_fd(fd_name, obj_name, self.tomasys)
+                reset_objective_status(obj_name, self.tomasys)
+                return fd_name
             return None
 
     # the DiagnosticStatus message process contains, per field
@@ -243,7 +246,8 @@ class Reasoner:
         self.tomasys.save(file="error.owl", format="rdfxml")
         sys.exit(0)
 
-    def handle_updatable_objectives(self, obj_in_error):
+    def handle_updatable_objectives(self, obj: str) -> None:
+        obj_in_error = self.get_objective_from_objective_id(obj)
         with self.ontology_lock:
             if obj_in_error.o_status == "UPDATABLE":
                 self.logger.info(
@@ -256,20 +260,9 @@ class Reasoner:
                                 comp_inst.name, comp_inst.c_status))
                         comp_inst.c_status = None
 
-    # selects configurations requested by the user
-    def select_requested_configurations(self):
-        objectives = self.search_objectives()
-        requested_configurations = dict()
-        for objective in objectives:
-            function = objective.typeF.name
-            if function in self.requested_configurations:
-                requested_configurations[objective] = \
-                    self.requested_configurations[function]
-        self.requested_configurations = dict()
-        return requested_configurations
-
     # find best fds for all objectives
-    def select_desired_configuration(self, obj_in_error):
+    def select_desired_configuration(self, obj: str) -> dict[str, str]:
+        obj_in_error = self.get_objective_from_objective_id(obj)
         with self.ontology_lock:
             self.logger.info(" >> Reasoner searches an FD ")
             desired_configuration = obtain_best_function_design(
@@ -301,41 +294,41 @@ class Reasoner:
             return objectives_in_error
 
         # EVALUATE functional hierarchy (objectives statuses) (MAPE - Analysis)
-        objectives_in_error = self.get_objectives_in_error()
+        objectives_in_error, o_status = self.get_objectives_in_error()
         if objectives_in_error == []:
             self.logger.info(
                 ">> No Objectives in ERROR: no adaptation is needed")
-        else:
-            for obj_in_error in objectives_in_error:
-                self.logger.warning(
-                    "Objective {0} in status: {1}".format(
-                        obj_in_error.name, obj_in_error.o_status))
+            return objectives_in_error
+
+        for obj_in_error, status in zip(objectives_in_error, o_status):
+            self.logger.warning(
+                "Objective {0} in status: {1}".format(
+                    obj_in_error, status))
         return objectives_in_error
 
     # MAPE-K: Plan step
-    def plan(self, objectives_in_error):
+    def plan(self, objectives_in_error: list[str]) -> dict[str, str]:
         if self.has_objective() is False or objectives_in_error == []:
             return dict()
 
         self.logger.info('  >> Started MAPE-K ** PLAN adaptation **')
-
-        desired_configurations = self.select_requested_configurations()
+        desired_configurations = dict()
         for obj_in_error in objectives_in_error:
             self.handle_updatable_objectives(obj_in_error)
 
-            if obj_in_error not in desired_configurations:
-                desired_config = self.select_desired_configuration(
-                    obj_in_error)
-                if desired_config is not None:
-                    desired_configurations[obj_in_error] = desired_config
+            desired_config = self.select_desired_configuration(
+                obj_in_error)
+            if desired_config is not None:
+                desired_configurations[obj_in_error] = desired_config
         return desired_configurations
 
     # MAPE-K: Execute step
-    def execute(self, desired_configurations):
+    def execute(self, desired_configurations: dict[str, str]):
         if self.has_objective() is False or desired_configurations == dict():
             return
 
         self.logger.info('  >> Started MAPE-K ** EXECUTION **')
         for objective in desired_configurations:
-            self.set_new_grounding(
-                desired_configurations[objective], objective)
+            if self.get_objective_from_objective_id(objective) is not None:
+                self.set_new_grounding(
+                    desired_configurations[objective], objective)

--- a/mros2_reasoner/mros2_reasoner/reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/reasoner.py
@@ -275,9 +275,10 @@ class Reasoner:
             return desired_configuration
 
     # MAPE-K: Analyze step
-    def analyze(self):
+    def analyze(self) -> list[str]:
         # PRINT system status
-        print_ontology_status(self.tomasys)
+        with self.ontology_lock:
+            print_ontology_status(self.tomasys)
 
         objectives_in_error = []
         if self.has_objective() is False:
@@ -288,10 +289,11 @@ class Reasoner:
 
         # EXEC REASONING to update ontology with inferences
         if self.perform_reasoning() is False:
-            self.logger.error('>> Reasoning error')
-            self.tomasys.save(
-                file="error_reasoning.owl", format="rdfxml")
-            return objectives_in_error
+            with self.ontology_lock:
+                self.logger.error('>> Reasoning error')
+                self.tomasys.save(
+                    file="error_reasoning.owl", format="rdfxml")
+                return objectives_in_error
 
         # EVALUATE functional hierarchy (objectives statuses) (MAPE - Analysis)
         objectives_in_error, o_status = self.get_objectives_in_error()

--- a/mros2_reasoner/mros2_reasoner/reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/reasoner.py
@@ -231,7 +231,9 @@ class Reasoner:
                 try:
                     sync_reasoner_pellet(
                         infer_property_values=True,
-                        infer_data_property_values=True)
+                        infer_data_property_values=True,
+                        debug=0
+                    )
                 except Exception as err:
                     self.logger.error(
                         "Error in perform_reasoning: {0}".format(err))

--- a/mros2_reasoner/mros2_reasoner/ros_reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/ros_reasoner.py
@@ -147,7 +147,7 @@ class RosReasoner(Node, Reasoner):
                     feedback_msg.qos_status.objective_type = \
                         str(request_objective.qos_expected.objective_type)
 
-                    fg_instance = self.tomasys.search_one(solvesO=objective)
+                    fg_instance = self.get_fg_solves_objective(objective)
                     if fg_instance is not None:
                         feedback_msg.qos_status.selected_mode = \
                             fg_instance.typeFD.name
@@ -179,7 +179,8 @@ class RosReasoner(Node, Reasoner):
             new_nfr = self.get_new_tomasys_nfr(
                  nfr_id, nfr_key.key, float(nfr_key.value))
             self.logger.info('Adding NFRs {}'.format(new_nfr))
-            new_objective.hasNFR.append(new_nfr)
+            new_objective = self.append_nfr_to_objective(
+                new_objective, new_nfr)
 
         # TODO: this is not working
         if not goal_request.qos_expected.selected_mode:
@@ -188,7 +189,8 @@ class RosReasoner(Node, Reasoner):
             self.set_initial_fd(goal_request.qos_expected.selected_mode)
 
         # TODO: shouldn't this be a swrl rule instead of hardcoded?
-        new_objective.o_status = 'UNGROUNDED'
+        new_objective = self.update_objective_status(
+            new_objective, 'UNGROUNDED')
 
         return True
 
@@ -226,14 +228,14 @@ class RosReasoner(Node, Reasoner):
                                 diagnostic_status.values[0].value))
                     else:
                         self.logger.warning(
-                            'Unsupported CS Message received: %s ' + str(
+                            'Unsupported CS Message received: {} '.format(
                                 diagnostic_status.values[0].key))
 
                 elif diagnostic_status.message == "QA status":
                     up_qa = self.update_qa(diagnostic_status)
                     if not up_qa:
                         self.logger.warning(
-                            'Unsupported QA TYPE received: %s ' + str(
+                            'Unsupported QA TYPE received: {} '.format(
                                 diagnostic_status.values[0].key))
 
     def request_configuration(self, desired_configuration, function_name):

--- a/mros2_reasoner/mros2_reasoner/ros_reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/ros_reasoner.py
@@ -1,6 +1,6 @@
 from rclpy.action import ActionServer
 from rclpy.action import CancelResponse
-from rclpy.callback_groups import ReentrantCallbackGroup
+from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
 from rclpy.node import Node
 from rclpy.parameter import Parameter
 
@@ -42,8 +42,6 @@ class RosReasoner(Node, Reasoner):
 
         self.is_initialized = False
 
-        self.cb_group = ReentrantCallbackGroup()
-
         # Start interfaces
         # subscriptions for different message types (named, pins, angle)
         # Node's default callback group is mutually exclusive. This would
@@ -54,7 +52,7 @@ class RosReasoner(Node, Reasoner):
             '/diagnostics',
             self.diagnostics_callback,
             1,
-            callback_group=self.cb_group)
+            callback_group=MutuallyExclusiveCallbackGroup())
 
         # Create action server
         self.objective_action_server = ActionServer(
@@ -62,7 +60,7 @@ class RosReasoner(Node, Reasoner):
             ControlQos,
             '/mros/objective',
             self.objective_action_callback,
-            callback_group=self.cb_group,
+            callback_group=MutuallyExclusiveCallbackGroup(),
             cancel_callback=self.objective_cancel_goal_callback)
 
         # Get desired_configuration_name from parameters
@@ -74,7 +72,7 @@ class RosReasoner(Node, Reasoner):
         self.metacontrol_loop_timer = self.create_timer(
             timer_period,
             self.metacontrol_loop_callback,
-            callback_group=self.cb_group)
+            callback_group=MutuallyExclusiveCallbackGroup())
 
         self.logger = self.get_logger()
 
@@ -246,7 +244,7 @@ class RosReasoner(Node, Reasoner):
         mode_change_cli = self.create_client(
                 MetacontrolFD,
                 '/mros/request_configuration',
-                callback_group=self.cb_group)
+                callback_group=MutuallyExclusiveCallbackGroup())
 
         while not mode_change_cli.wait_for_service(timeout_sec=1.0):
             self.logger.warning(

--- a/mros2_reasoner/mros2_reasoner/ros_reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/ros_reasoner.py
@@ -54,13 +54,14 @@ class RosReasoner(Node, Reasoner):
             1,
             callback_group=MutuallyExclusiveCallbackGroup())
 
+        cb_group = MutuallyExclusiveCallbackGroup()
         # Create action server
         self.objective_action_server = ActionServer(
             self,
             ControlQos,
             '/mros/objective',
             self.objective_action_callback,
-            callback_group=MutuallyExclusiveCallbackGroup(),
+            callback_group=cb_group,
             cancel_callback=self.objective_cancel_goal_callback)
 
         # Get desired_configuration_name from parameters
@@ -72,7 +73,7 @@ class RosReasoner(Node, Reasoner):
         self.metacontrol_loop_timer = self.create_timer(
             timer_period,
             self.metacontrol_loop_callback,
-            callback_group=MutuallyExclusiveCallbackGroup())
+            callback_group=cb_group)
 
         self.logger = self.get_logger()
 

--- a/mros2_reasoner/mros2_reasoner/ros_reasoner.py
+++ b/mros2_reasoner/mros2_reasoner/ros_reasoner.py
@@ -97,13 +97,14 @@ class RosReasoner(Node, Reasoner):
             function_name = self.get_function_name_from_objective_id(
                 cancel_request.request.qos_expected.objective_id)
 
-            if function_name is not None:
-                reconfiguration_result = self.request_configuration(
-                    'fd_unground',
-                    function_name)
-            else:
-                reconfiguration_result = None
+            if function_name is None:
+                self.logger.info('Objective {} cancel req failed'.format(
+                    cancel_request.request.qos_expected.objective_id))
+                return CancelResponse.REJECT
 
+            reconfiguration_result = self.request_configuration(
+                'fd_unground',
+                function_name)
             if reconfiguration_result is None \
                or reconfiguration_result.success is False:
                 self.logger.info('Objective {} cancel req failed'.format(
@@ -271,11 +272,16 @@ class RosReasoner(Node, Reasoner):
                 'desired_configurations are: {}'.format(
                     desired_configurations))
         for objective in desired_configurations:
-            reconfiguration_result = self.request_configuration(
-                desired_configurations[objective], str(objective.typeF.name))
+            if self.get_objective_from_objective_id(objective) is not None:
+                reconfiguration_result = self.request_configuration(
+                    desired_configurations[objective],
+                    self.get_function_name_from_objective_id(objective))
 
-            if reconfiguration_result is not None \
-               and reconfiguration_result.success is True:
+                if reconfiguration_result is None \
+                   or reconfiguration_result.success is False:
+                    self.logger.error('= RECONFIGURATION FAILED =')
+                    continue
+
                 self.logger.info(
                     'Got Reconfiguration result {}'.format(
                         reconfiguration_result.success))
@@ -283,9 +289,6 @@ class RosReasoner(Node, Reasoner):
                 # Process adaptation feedback to update KB:
                 self.set_new_grounding(
                     desired_configurations[objective], objective)
-            else:
-                self.logger.error('= RECONFIGURATION FAILED =')
-                return
 
     # main metacontrol loop
     def metacontrol_loop_callback(self):

--- a/mros2_reasoner/mros2_reasoner/tomasys.py
+++ b/mros2_reasoner/mros2_reasoner/tomasys.py
@@ -10,6 +10,7 @@
 
 from owlready2 import destroy_entity
 from owlready2 import get_ontology
+from typing import Tuple
 import logging
 
 
@@ -65,10 +66,13 @@ def read_ontology_file(ontology_file_array):
 
 # To reset the individuals that no longer hold due to adaptation
 # for the moment, only Objective individuals statuses
-# - tomasys: ontology holding the Tbox
-def reset_objective_status(objective, status=None):
+# - ontology: ontology holding the Tbox
+def reset_objective_status(objective_name: str, ontology, status=None):
     # logging.warning("\nReseting obj {0}".format(objective.name))
-    objective.o_status = status
+    objective = ontology.search_one(
+        iri="*{}".format(objective_name), is_a=ontology.Objective)
+    if objective is not None:
+        objective.o_status = status
 
 
 def reset_fd_realisability(ontology, c_name):
@@ -143,16 +147,18 @@ def update_fg_measured_qa(fg, measured_qa):
 
 # Evaluates the Objective individuals in the KB and returns a list with
 # those in error
-def get_objectives_in_error(objectives):
+def get_objectives_in_error(objectives) -> Tuple[list[str], list[str]]:
     objectives_internal_error = []
+    objectives_internal_error_status = []
     for o in objectives:
         if o.o_status in ["UNGROUNDED",
                           "UPDATABLE",
                           "IN_ERROR_FR",
                           "IN_ERROR_NFR",
                           "IN_ERROR_COMPONENT"]:
-            objectives_internal_error.append(o)
-    return objectives_internal_error
+            objectives_internal_error.append(str(o.name))
+            objectives_internal_error_status.append(str(o.o_status))
+    return objectives_internal_error, objectives_internal_error_status
 
 
 def get_function_grounding(o, ontology):
@@ -217,30 +223,37 @@ def obtain_best_function_design(o, ontology):
         return None
 
 
-def ground_fd(fd, objective, ontology):
+def ground_fd(fd_name: str, objective_name: str, ontology):
     """Given a FunctionDesign fd and an Objective objective, creates an
        individual FunctionGrounds with typeF fd and solve) objective returns
        the fg
     """
-    fg = ontology.FunctionGrounding(
-        "fg_" +
-        fd.name.replace(
-            'fd_',
-            ''),
-        namespace=ontology,
-        typeFD=fd,
-        solvesO=objective)
-    # TODO: ground objectives required by FD
-    return fg
+    objective = ontology.search_one(
+        iri="*{}".format(objective_name), is_a=ontology.Objective)
+    fd = ontology.search_one(
+        iri="*{}".format(fd_name), is_a=ontology.FunctionDesign)
+    if objective is not None and fd is not None:
+        fg = ontology.FunctionGrounding(
+            "fg_" +
+            fd.name.replace(
+                'fd_',
+                ''),
+            namespace=ontology,
+            typeFD=fd,
+            solvesO=objective)
+        # TODO: ground objectives required by FD
+        return fg
 
 
-def remove_objective_grounding(objective, ontology):
+def remove_objective_grounding(objective_name, ontology):
     """Given an objective individual, removes the grounded hierarchy (fg tree)
         that solves it.
     """
-    fg = ontology.search_one(solvesO=objective)
-    if fg:
-        destroy_entity(fg)
+    objective = ontology.search_one(iri="*{}".format(objective_name))
+    if objective is not None:
+        fg = ontology.search_one(solvesO=objective)
+        if fg is not None:
+            destroy_entity(fg)
 
 
 def get_measured_qa(key, ontology):

